src/lane_keep_assist/scripts/lane_keep.py

```diff
@@ -0,0 +1,107 @@
+        self.steering_pub_test = rospy.Publisher('lane_info', LaneStatus, queue_size=10)
+        self.gradient_state = 1
+        self.hls_state = 1
+        self.segmentation_state = 1
+        self.gradient_lanes = []
+        self.hls_lanes = []
+        self.segmentation_lane = []
+        self.gradient_percent_diff = 0
+        self.hls_percent_diff = 0
+        self.image = None
+        self.segmented_image = None
+        self.road_colour = 0
+        self.first_detection = True
+
+    def listener(self):
+        rospy.init_node('LaneKeepAssist', anonymous=True)
+        rospy.Subscriber("airsim/image_raw", Image, self.handle_camera_data)
+        rospy.Subscriber("segmented_image", Image, self.handle_segmented_image)
+
+        rate = rospy.Rate(10)
+        while not rospy.is_shutdown():
+            if self.image is not None and self.segmented_image is not None:
+                self.process()
+
+                # Lane status message carries info on the lane boundaries, road boundaries and the percent diff
+                lane_msg = LaneStatus()
+                lane_msg.lane_gradient_status = self.gradient_state
+                lane_msg.lane_hls_status = self.hls_state
+                lane_msg.lane_segmentation_status = self.segmentation_state
+                lane_msg.gradient_diff = self.gradient_percent_diff
+                lane_msg.hls_diff = self.hls_percent_diff
+                lane_msg.gradient_lane_bounds = []
+                lane_msg.hls_lane_bounds = []
+                lane_msg.segmentation_lane_bounds = []
+
+                for lane_result in self.gradient_lanes:
+                    lane = LaneLine()
+                    lane.x1, lane.y1, lane.x2, lane.y2 = lane_result
+                    lane_msg.gradient_lane_bounds.append(lane)
+
+                for lane_result in self.hls_lanes:
+                    lane = LaneLine()
+                    lane.x1, lane.y1, lane.x2, lane.y2 = lane_result
+                    lane_msg.hls_lane_bounds.append(lane)
+
+                for lane_result in self.segmentation_lane:
+                    lane = LaneLine()
+                    lane.x1, lane.y1, lane.x2, lane.y2 = lane_result
+                    lane_msg.segmentation_lane_bounds.append(lane)
+
+                self.steering_pub_test.publish(lane_msg)
+            rate.sleep()
+
+    # Process the normal scene image and road boundaries image
+    def process(self):
+        # Unpacking the results
+        lane_results, percent_differences, lanes, road_colour = lane_turn.lane_interpret(self.image, self.segmented_image)
+        self.gradient_state, self.hls_state, self.segmentation_state = lane_results
+        self.gradient_percent_diff, self.hls_percent_diff = percent_differences
+        self.gradient_lanes, self.hls_lanes, self.segmentation_lane = lanes
+
+        # Save the first colour detected for the road detection
+        if self.first_detection:
+            self.road_colour = road_colour
```

Comment:
Is this not a constant?

Reply:
No, the way this works is that I just detect the primary colour in the bottom quarter to the screen which is most likely the road. But if the car goes way off the road that detection could change. This just locks it to what was recorded at launch assuming that car starts on the road. If it goes off the road the colour changes and any lane detections will be ignnored.